merkle-tree/reference/src/lib.rs

```diff
@@ -156,4 +156,40 @@ where
     pub fn get_leaf_index(&self, leaf: &[u8; 32]) -> Option<usize> {
         self.layers[0].iter().position(|node| node == leaf)
     }
+
+    pub fn verify(
+        &self,
+        leaf: &[u8; 32],
+        proof: &BoundedVec<[u8; 32]>,
+        leaf_index: usize,
+    ) -> Result<bool, ReferenceMerkleTreeError> {
+        if leaf_index >= self.capacity {
+            return Err(ReferenceMerkleTreeError::LeafDoesNotExist(leaf_index));
+        }
+
+        let mut computed_hash = *leaf;
+        let mut current_index = leaf_index;
+
+        for (i, sibling_hash) in proof.iter().enumerate() {
+            let is_left = current_index % 2 == 0;
+            let hashes = if is_left {
+                [&computed_hash[..], &sibling_hash[..]]
+            } else {
+                [&sibling_hash[..], &computed_hash[..]]
+            };
+
+            computed_hash = H::hashv(&hashes)?;
+
+            // Move to the parent index for the next iteration
+            current_index /= 2;
+
+            // If the proof contains more levels than needed, we might compute the wrong hash
```

Comment:
1. Shouldn't we return an explicit error in such case? A silent `break` and just trimming the proof will make mistakes harder to debug, reminds me of `from_bytes` modulo in arkworks. :joy: 
2. In case we use canopy, we should keep it in mind when verifying the size (`self.height - self.canopy`). I'm fine with accepting both full proofs and partial proofs, but it would be still great to check the size explicitly and return an error.

Reply:
good point, since we don't have an update proof method in this tree I will just throw an error if proof length != height.

Reply:
done